In [1]:
import pandas as pd
iris = pd.read_csv('iris.csv')

In [2]:
iris.sample(10)

sepal_length  sepal_width  petal_length  petal_width     species
51            6.4          3.2           4.5          1.5  versicolor
91            6.1          3.0           4.6          1.4  versicolor
29            4.7          3.2           1.6          0.2      setosa
142           5.8          2.7           5.1          1.9   virginica
45            4.8          3.0           1.4          0.3      setosa
63            6.1          2.9           4.7          1.4  versicolor
139           6.9          3.1           5.4          2.1   virginica
40            5.0          3.5           1.3          0.3      setosa
20            5.4          3.4           1.7          0.2      setosa
7             5.0          3.4           1.5          0.2      setosa

### 假设检验
根据样本数据，推测总体平均值的大小   
根据中心极限定理，sepal_length平均值的分布是正态分布，且数据量大于30，进行z检验   

In [5]:
# 样本平均值
import numpy as np
np.mean(iris['sepal_length'])

5.843333333333334

零假设为sepal_length的平均数为5.8，备择假设为不是5.8

In [4]:
import statsmodels.stats.weightstats
statsmodels.stats.weightstats.ztest(iris['sepal_length'], value=5.8)

(0.6409183514112012, 0.5215757321117633)

P-value的值是0.5216，大于0.05，可以看到整体平均值为5.8的概率为52.16%，远远大于5%，所以不能拒绝零假设，猜测整体平均数是5.8

零假设为sepal_length的平均数为5.7，备择假设为不是5.7

In [6]:
statsmodels.stats.weightstats.ztest(iris['sepal_length'], value=5.7)

(2.1199607008216415, 0.03400935962004441)

P-value为0.034, 小于0.05，拒绝零假设，承认备择假设，平均数不是5.7

比较样本中两种类型的花的平均值是否一致

In [7]:
iris_setosa = iris[iris['species'] == 'setosa']
iris_virginica = iris[iris['species'] == 'virginica']

零假设为两种花的sepal_length平均值一样，备择假设为不一样

In [8]:
import scipy.stats
scipy.stats.ttest_ind(iris_setosa['sepal_length'], iris_virginica['sepal_length'])

Ttest_indResult(statistic=-15.386195820079404, pvalue=6.892546060674059e-28)

P-value非常小，说明平均值存在显著的区别，可以通过sepal_length来区分这两种花   

In [9]:
print(np.mean(iris_setosa['sepal_length']), np.mean(iris_virginica['sepal_length']))
# 样本的平均值也是非常大

5.006 6.587999999999998
